### ISSUE TITLE
Fetch returning NotFoundError even when state is set

### ISSUE THREAD

**Issue by FoxxMD** [jhollingworth/marty#97]
First off, fantastic flux implementation!

I am trying to do a simple fetch using localStorage but my store always returns `NotFoundError` even though I can see the object set in the state (by calling `getState` or just checking `this.state['myobj']`. What am I doing wrong??

**Store**
```
var AuthStore = Marty.createStore({
    displayName:'Auth',
    handlers: {
        updateCredentials: AuthConstants.UPDATE_CREDENTIALS,
        clearCredentials: AuthConstants.CLEAR_CREDENTIALS
    },
    getInitialState: function() {
      return {};
    },
    getCredentials: function(){
        return this.fetch({
            id:'credentials',
            locally: function(){
                return this.state['credentials'];
            },
            remotely: function(){
                return credSource.getCreds();
            }
        })
    },
    updateCredentials: function(payload) {
        this.state['credentials'] = payload;
        credSource.setCreds(payload);
        this.hasChanged();
    },
    clearCredentials: function(payload) {
        this.state['credentials'] = undefined;
        credSource.setCreds(undefined);
        this.hasChanged();
    }
});
module.exports = AuthStore;
```

**State Source**

```
var credSource = Marty.createStateSource({
    type:'jsonStorage',
    storage: window.localStorage,
    getCreds: function(){
       return AuthSourceActionCreators.updateCredentials(this.get('credentials'));
    },
    setCreds: function(creds){
        this.set('credentials', creds);
    }
});
module.exports = credSource;
```

**SourceActionCreator**

```
var AuthSourceActionsCreators = Marty.createActionCreators({
    updateCredentials: AuthConstants.UPDATE_CREDENTIALS()
});
module.exports = AuthSourceActionsCreators;
```

Thanks for looking! I feel like I am overlooking something fairly simple but can't seem to find it...

**Comment by jhollingworth** [comment]
Hey, thanks for giving Marty a try :) Here's a [working fiddle based on your example](http://jsfiddle.net/L0yr9a3q/). Let me know if it makes sense or you've got questions

**Comment by FoxxMD** [comment]
Your example was perfect! I think my mistakes was not assigning `getCreds` to the state in my store after retrieving it with remotely.

I have a conceptual question.
You moved setting and clearing credentials *in storage* into the source action creator instead of in the auth store -- do you trying to restrict manipulating source data, whether http or storage, to actions in source action creator? From the examples it seemed like it was only used as a workaround for cyclic dependencies with require.

Basically what I'm seeing is

**Stores**
* State manipulation
* Fetching data

**Action Creators**
* Business logic for app
* Invoke source action creators (if source data)
* Invoke handlers on store (if no source data)

**Source Action Creators**
* Manipulate source data
* Invoke handlers on store

Is this a good way to look at logic? I'm trying to wrap my head around where to use what.

**Comment by oliverwoodings** [comment]
I think you are missing a few entities there - add in views and ensure the concept of an 'action' exists then it all makes sense:

**Action**
* Description of an event that has occurred somewhere in the domain of the app
* Can have data attached to it

**Stores**
* Keep local state
* Request remote state from sources
* Update local state based on actions

**Sources**
* Responsible for storing and retrieving data
* Invoke Source Action Creators to tell stores that remote state has been loaded

**Views**
* Render the application
* Retrieve state from stores
* Invoke Action Creators

**Action Creators**
* Dispatch local actions (typically triggered by views), often to trigger state updates in stores

**Source Action Creators**
* Dispatch source actions (typically triggered by stores loading data)

**Comment by jhollingworth** [comment]
If you haven't read it yet, the [Fetching data](http://martyjs.org/guides/stores/fetching-data.html) guide might be helpful.

> Should Source Action Creators only be dispatched after the source has finished performing it's action, or can they be used to trigger source actions

Source Action Creators only exist to get around cyclic dependencies. They should only be invoked by State sources.

The previous fiddle was actually cheating with regards to getting the state from the Cred source. Any new state coming from a source should go through a source action creator so that all stores have the opportunity to do something with that state. This [revised fiddle](http://jsfiddle.net/p0d08k3x/) is how you should be doing it

**Comment by FoxxMD** [comment]
Thanks @jhollingworth . So I should implement my own promises when using local storage. :+1:

**Comment by jhollingworth** [comment]
no problem!
